src/arrayformat.rs

```diff
@@ -11,9 +11,203 @@ use super::{
+            (Some(_), None) => ArrayDisplayMode::VSplit,
+            (None, _) => ArrayDisplayMode::Full,
+        }
+    }
+
+    fn h_split_offset(&self) -> Option<Ix> {
+        match self {
+            ArrayDisplayMode::DoubleSplit(axis) | ArrayDisplayMode::HSplit(axis) => {
+                Some(axis + 1usize)
+            },
+            _ => None
+        }
+    }
+}
+
+fn format_array_v2<A, S, D, F>(view: &ArrayBase<S, D>,
+                                  f: &mut fmt::Formatter,
+                                  mut format: F,
+                                  limit: usize) -> fmt::Result
+    where F: FnMut(&A, &mut fmt::Formatter) -> fmt::Result,
+          D: Dimension,
+          S: Data<Elem=A>,
+{
+    let display_mode = ArrayDisplayMode::from_array(view, limit);
+
+    let ndim = view.dim().into_dimension().slice().len();
+    let nth_idx_max = if ndim > 0 { Some(view.shape().iter().last().unwrap()) } else { None };
+
+    // None will be an empty iter.
+    let mut last_index = match view.dim().into_dimension().first_index() {
+        None => view.dim().into_dimension().clone(),
+        Some(ix) => ix,
+    };
+    for _ in 0..ndim {
+        write!(f, "[")?;
+    }
+    let mut first = true;
+    // Shows if ellipses for vertical split were printed.
+    let mut printed_ellipses_v = false;
+    // Shows if ellipses for horizontal split were printed.
+    let mut printed_ellipses_h = false;
+    // Shows if the row was printed for the first time after horizontal split.
+    let mut no_rows_after_skip_yet = false;
+
+    // Simply use the indexed iterator, and take the index wraparounds
+    // as cues for when to add []'s and how many to add.
+    for (index, elt) in view.indexed_iter() {
+        let index = index.into_dimension();
+        let take_n = if ndim == 0 { 1 } else { ndim - 1 };
+        let mut update_index = false;
+
+        let mut print_row = true;
+        match display_mode {
+            ArrayDisplayMode::HSplit(axis) | ArrayDisplayMode::DoubleSplit(axis) => {
+                let sa_idx_max = view.shape().iter().skip(axis).next().unwrap();
+                let sa_idx_val = index.slice().iter().skip(axis).next().unwrap();
+                if sa_idx_val >= &limit && sa_idx_val < &(sa_idx_max - &limit) {
+                    print_row = false;
+                    no_rows_after_skip_yet = true;
+                }
+            },
+            _ => {}
+        }
+
+        for (i, (a, b)) in index.slice()
+            .iter()
+            .take(take_n)
+            .zip(last_index.slice().iter())
+            .enumerate() {
+            if a != b {
+                if print_row {
+                    printed_ellipses_v = false;
+                    // New row.
+                    // # of ['s needed
+                    let n = ndim - i - 1;
+                    if !no_rows_after_skip_yet {
+                        for _ in 0..n {
+                            write!(f, "]")?;
```

Comment:
[`std::str::repeat`](https://doc.rust-lang.org/std/primitive.str.html#method.repeat) could be handy here - it should allow you to get rid of a bunch of for cycles.